ankidroid/Anki-Android
# Avoid unhandled out-of-space and OOM on large imports

## Pull Request template

## Purpose / Description
This is a continuation of #4622 from @camparijet - his original idea - do partial commits to the database while importing - seemed correct but I could not get it to work at the time.

Now with very large decks being more normal (#6419 #6438 #6435) I was reminded about this and wanted to try it again

## Fixes
Fixes #6419 at least in part
Maybe fixes #4481 #2596

## Approach
1. Scan a zip before even trying it, tell the user if there is not enough space to decompress
1. During import, commit work and clear scratch space periodically so we don't run out of memory
1. During import if we run out of space, report that to the user nicely
1. Post import, during cleanup, if we run out of space report that to the user nicely

## How Has This Been Tested?

This was tested on an emulator, with each scenario repeating these steps

1. shutdown the emulator
1. wipe data
1. set the external sdcard size to a specific size
1. start the emulator
1. upload the attached OOM collection to /sdcard/Download/
1. import the OOM collection (original https://drive.google.com/open?id=0B5kybuti0pOIQXoxRmtHMWc2c00 and I will try to attach it here 
[OOMDeck.apkg.gz](https://github.com/ankidroid/Anki-Android/files/4775794/OOMDeck.apkg.gz)
)

Here are the scenarios to run, by size of sdcard, and what you should see:

1. 350MB sdcard, you should get an error that there isn't enough space to unzip
1. 500MB of sdcard, you should get an error during import that import failed because disk was full
1. 1GB of sdcard, message that import succeeded but cleanup failed. Deck works. DB check afterwards works.
1. 1.5GB of sdcard, everything should work.

Without this patch, it just fails everywhere.

## Learning (optional, can help others)
Look really closely at PRs before you discard them? This one was one line away from working last time. (it was missing `revlog.clear()`


## Checklist
